File: backend/app/models/product.rb
class Product < ApplicationRecord
  mount_uploader :image, ImageUploader

  validates :name, :caption, :price, presence: true
  validates :price, numericality: { greater_than: 0 }

  has_many :likes, dependent: :destroy
  has_many :reviews, dependent: :destroy

  scope :ranking, ->(count) { all.order('likes_count desc').limit(count) }
  scope :limit_index, -> { limit(INDEX_NUM) }
  scope :page_offset, -> (page_num) {offset(INDEX_NUM * page_num)}

  def self.index_pagenation(page)
    limit_index.page_offset(page)
  end

  def rate_average_sum
    return unless reviews.average(:rate)

    (reviews.average(:rate) * 2).floor / 2.to_f
  end

  def rate_average
    (rate_sum.to_f / reviews_count).floor(1)
  end

  def api_json
    {
      id: id,
      name: name,
      url: url,
      shopname: shopname,
      price: price,
      caption: caption,
      image: { url: image.url },
      reviews_count: reviews_count,
      rate_average: rate_average
    }
  end
end

File: frontend/src/mock/product.ts
import image1 from 'images/product-01.jpg';
import image2 from 'images/product-02.jpg';
import image3 from 'images/product-03.jpg';
import { Product } from 'model/index';

export const products: Product[] = [
  {
    id: 1,
    name: 'コーヒー器具の名前1',
    image: { url: image1 },
    shopname: 'コーヒーのショップの名前1',
    price: 100,
    url: 'http://www.example.com/1',
    caption: 'アイテムの説明が入ります。1',
    likes_count: 1,
  },
  {
    id: 2,
    name: 'コーヒー器具の名前2',
    image: { url: image2 },
    shopname: 'コーヒーのショップの名前2',
    price: 200,
    url: 'http://www.example.com/2',
    caption: 'アイテムの説明が入ります。2',
    likes_count: 2,
  },
  {
    id: 3,
    name: 'コーヒー器具の名前3',
    image: { url: image3 },
    shopname: 'コーヒーのショップの名前3',
    price: 300,
    url: 'http://www.example.com/3',
    caption: 'アイテムの説明が入ります。3',
    likes_count: 3,
  },
];

export default products;
